File: Design-Patterns/Visitor/Cpp/weapon.cpp
#include <iostream>
#include "equipment.h"
#include "equipment_visitor.h"

weapon::weapon(std::string name, int loading) : equipment(name, loading)
{

}

void weapon::run_visitor(std::shared_ptr<equipment_visitor> visitor)
{
	visitor->visit_weapon(std::shared_ptr<weapon>(std::shared_ptr<weapon>(), this));
}

void weapon::attack()
{
	std::cout << "Weapon " + name_ + " attacks." << std::endl;
}

File: Design-Patterns/Visitor/CSharp/Equipments/Weapon.cs
using System;
using Visitor.Visitors;

namespace Visitor.Equipments
{
    public class Weapon : Equipment
    {
        public Weapon(string name, int loading) : base(name, loading)
        {

        }

        public override void RunVisitor(EquipmentVisitor visitor)
        {
            visitor.VisitWeapon(this);
        }

        public void Attack()
        {
            Console.WriteLine($"Weapon {Name} attacks.");
        }
    }
}
